ComplianceAsCode/content
# Adds utilities for OSCAL Component Definitions from policies

#### Description:

This PR adds utilities under the `utils` directory to create OSCAL component definitions using the files under the `controls` directory to populate the implemented requirements and mapping to the rules.

##### What was added?

- Core logic under `utils/oscal`
- Unit tests under `test/unit/utils/oscal`
- Documentation for the utilities

#### Rationale:

The policy was used to initially fill in the control responses because it contains all the required information for a component definition (status, description, rule mapping). However, the class `ComponentDefinitionGenerator` takes a `ControlSelector` class in order to allow how control response information is selected and constructed separate from the generation logic.

I tried to follow the guidance of one-step one commit to keep the options open for breaking this down into smaller PRs. However, all the changes would be needed to get an OSCAL component definition that would be compatible with `compliance-trestle`.

- Fixes #11106 

#### Review Hints:

This is PR is part two to #11286 so please see that PR to review the first two commits. **Merged**

##### Important Notes

- To support installing `compliance-trestle` during CI job execution for testing, changes were made to the workflow files in the following commits:
https://github.com/ComplianceAsCode/content/pull/11292/commits/9fea72dbd06a1042759b788d184fd5a2fe8b7c68
https://github.com/ComplianceAsCode/content/pull/11292/commits/5fa74a9d82419f44ec47ebfd9f145378e33a5151


- It performs a mapping from SSG status to OSCAL implementation status in the `OscalStatus` class
- The script uses the "Section *letter*:" convention in the control notes to create statements under the implemented requirements.
- The script maps parameter to rules uses the `xccdf_variable` field under `template.vars`
- To determine what responses will mapped to the controls in the OSCAL profile the control id and label property from the resolved catalog is searched.

##### How to test

An example of how to execute the script:

```bash
$ ./build_product ocp4
$ ./utils/rule_dir_json.py
$ ./utils/oscal/build_cd_from_policy.py -o build/ocp4.json -p fedramp_rev4_high -pr ocp4 -c nist_ocp4:high
```
Example of how to verify compatibility with `compliance-trestle`
```bash
$ mkdir tmp-trestle-dir
$ cd tmp-trestle-dir && trestle init
$ trestle import -f https://raw.githubusercontent.com/GSA/fedramp-automation/master/dist/content/rev4/baselines/json/FedRAMP_rev4_HIGH-baseline_profile.json -o fedramp_rev4_high
$ trestle import -f ../content/build/ocp4.json -o ocp4
$ trestle author component-generate --name ocp4 -o markdown/components/ocp4
# Here you will notice that the implementation and rules are populated in the Markdown
$ cat markdown/components/ocp4/ocp4/fedramp_rev4_high/cm/cm-6.1.md
```
